### PR TITLE
LLDB pretty-printers test runner

### PR DESCRIPTION
**Pull request by ortem**
Implement the test runner for bundled LLDB pretty-printers.
Inspired by [rustc test runner](https://github.com/rust-lang/rust/blob/master/src/tools/compiletest/src/runtest.rs) and corresponding [debuginfo tests](https://github.com/rust-lang/rust/tree/master/src/test/debuginfo).

### CODE REVIEW

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

File: prettyPrinters/providers.py [hunk 1]
```diff
@@ -145,7 +149,9 @@ def StdStrSummaryProvider(valobj, dict):
     process = data_ptr.GetProcess()
     data = process.ReadMemory(start, length, error)
     if error.Success():
-        return '"%s"' % data
+        if system() == "Darwin":
+            return '"%s"' % data
+        return '"%s"' % data.decode(encoding='UTF-8')
```
Comment (Undin) [inline prettyPrinters/providers.py]: Look like it shouldn't be in this PR because it's part of #3124

File: build.gradle.kts [hunk 2]
```diff
@@ -131,15 +132,16 @@ allprojects {
     }
 }
 
-val channelSuffix = if (channel.isBlank()) "" else "-$channel"
+val channelSuffix = if (channel.isBlankKt()) "" else "-$channel"
```
Comment (Undin) [inline build.gradle.kts]: Let's revert all changes related to our experiments with jdk11

File: build.gradle.kts [hunk 3]
```diff
@@ -131,15 +132,16 @@ allprojects {
...
 
 project(":") {
     val clionVersion = prop("clionVersion")
+    val clionFullName = if (CI) "clion" else "clion-$clionVersion"
     val versionSuffix = "-$platformVersion$channelSuffix"
     version = "0.2.0.${prop("buildNumber")}$versionSuffix"
     intellij {
         pluginName = "intellij-rust"
-//        alternativeIdePath = "deps/clion-$clionVersion"
+        alternativeIdePath = "/Users/jetbrains/Library/Application Support/JetBrains/Toolbox/apps/CLion/ch-0/183.4284.156/CLion.app"
```
Comment (Undin) [inline build.gradle.kts]: Looks like it added by accident

File: pretty_printers_tests/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+use std::fs;
+use std::fs::read_dir;
+use std::path::Path;
+
+use test_runner::{LLDBConfig, LLDBTestRunner};
+
+fn main() -> Result<(), ()> {
+    let settings_file =
+        if cfg!(target_os = "linux") {
```
Comment (Undin) [inline pretty_printers_tests/src/main.rs]: Not sure but maybe separate cfg functions for all supported os are better?
Something like:
```rust
#[cfg(target_os = "linux")]
fn settings() -> &str {
    "Settings_linux.toml"
}
```

Comment (ortem) [reply pretty_printers_tests/src/main.rs]: What do you think about this?
```
#[cfg(target_os = "linux")]
static SETTINGS: &str = "Settings_linux.toml";
#[cfg(target_os = "macos")]
static SETTINGS: &str = "Settings_macos.toml";
#[cfg(all(not(target_os = "linux"), not(target_os = "macos")))]
panic!("Unsupported platform");

fn main() -> Result<(), ()> {
    let settings = fs::read_to_string(SETTINGS).expect(&format!("Cannot read {}", SETTINGS));
    let lldb_config: LLDBConfig = toml::from_str(&settings).expect(&format!("Invalid {}", SETTINGS));
```

Comment (Undin) [reply pretty_printers_tests/src/main.rs]: @ortem generally, looks ok
But it seems that `SETTINGS` can be conts

File: pretty_printers_tests/Settings_macos.toml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+test_dir = "./tests"
+pretty_printers_path = "./../prettyPrinters/"
+lldb_batchmode = "./lldb_batchmode.py"
+lldb_lookup = "lookup"
+lldb_python = "/Applications/Xcode.app/Contents/SharedFrameworks/LLDB.framework/Resources/Python"
```
Comment (Undin) [inline pretty_printers_tests/Settings_macos.toml]: Could you add `# BACKOMPAT: 2018.2` comment here to use python from clion distribution

File: build.gradle.kts [hunk 6]
```diff
@@ -179,14 +180,15 @@ project(":") {
...
-        dest(file("${project.projectDir}/deps/clion-$clionVersion.tar.gz"))
+        dest(file("${project.projectDir}/deps/$clionFullName.tar.gz"))
     }
     val unpackClion = task<Copy>("unpackClion") {
-        onlyIf { !file("${project.projectDir}/deps/clion-$clionVersion").exists() }
-        from(tarTree("deps/clion-$clionVersion.tar.gz"))
+        onlyIf { !file("${project.projectDir}/deps/$clionFullName").exists() }
+        from(tarTree("deps/$clionFullName.tar.gz"))
         into(file("${project.projectDir}/deps"))
+        file("${project.projectDir}/deps/clion-$clionVersion").renameTo(file("${project.projectDir}/deps/$clionFullName"))
```
Comment (Undin) [inline build.gradle.kts]: Why do we need this?

File: build.gradle.kts [hunk 7]
```diff
@@ -179,14 +180,15 @@ project(":") {
     }
 
     val downloadClion = task<Download>("downloadClion") {
-        onlyIf { !file("${project.projectDir}/deps/clion-$clionVersion.tar.gz").exists() }
+        onlyIf { !file("${project.projectDir}/deps/$clionFullName.tar.gz").exists() }
```
Comment (Undin) [inline build.gradle.kts]: Do we really need rename all intermediate artifacts?
Maybe it's enough to rename only final folder?

File: pretty_printers_tests/Settings_macos.toml (hunk 5 above)
Comment (Undin) [reply pretty_printers_tests/Settings_macos.toml]: I mean to add `# BACKOMPAT: 2018.2` with some useful message to understand what need todo here

**Comment by mchernyavsky** [comment]
bors retry

**Comment by Undin** [comment]
Looks like it really doesn't work on appveyor
@ortem could you check?

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

**Comment by ortem** [comment]
bors try

**Pull request merged**